File: service/src/main/java/no/nav/melosys/service/kontroll/feature/arbeidutland/kontroll/ArbeidUtlandKontroll.java
package no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll;

import java.util.ArrayList;
import java.util.List;

import no.nav.melosys.domain.mottatteopplysninger.MottatteOpplysningerData;
import no.nav.melosys.domain.mottatteopplysninger.data.ForetakUtland;
import no.nav.melosys.domain.mottatteopplysninger.data.arbeidssteder.FysiskArbeidssted;
import no.nav.melosys.domain.kodeverk.begrunnelser.Kontroll_begrunnelser;
import exception.KontrolldataFeilType;
import no.nav.melosys.service.kontroll.regler.AdresseRegler;
import no.nav.melosys.service.validering.Kontrollfeil;

public class ArbeidUtlandKontroll {
    static String ARBEIDSSTED_FIRMANAVN = "mottatteopplysninger.arbeidPaaLand.fysiskeArbeidssteder[%d].virksomhetNavn";
    static String ARBEIDSSTED_LAND = "mottatteopplysninger.arbeidPaaLand.fysiskeArbeidssteder[%d].adresse.landkode";
    static String FORETAK_UTLAND_NAVN = "mottatteopplysninger.foretakUtland[%d].navn";
    static String FORETAK_UTLAND_LAND = "mottatteopplysninger.foretakUtland[%d].adresse.landkode";

    public static Kontrollfeil arbeidsstedManglerFelter(MottatteOpplysningerData mottatteOpplysningerData) {
        List<FysiskArbeidssted> fysiskArbeidsstedListe = mottatteOpplysningerData.arbeidPaaLand.fysiskeArbeidssteder;

        List<String> felterMedFeil = new ArrayList<>();
        for (int i = 0; i < fysiskArbeidsstedListe.size(); i++) {
            FysiskArbeidssted fysiskArbeidssted = fysiskArbeidsstedListe.get(i);
            if (AdresseRegler.manglerArbeidsstedVirksomhetsnavn(fysiskArbeidssted)) {
                felterMedFeil.add(String.format(ARBEIDSSTED_FIRMANAVN, i));
            }
            if (AdresseRegler.manglerArbeidsstedLandkode(fysiskArbeidssted)) {
                felterMedFeil.add(String.format(ARBEIDSSTED_LAND, i));
            }
        }
        return felterMedFeil.isEmpty() ? null
            : new Kontrollfeil(Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSSTED, felterMedFeil, KontrolldataFeilType.FEIL);
    }

    public static Kontrollfeil foretakUtlandManglerFelter(MottatteOpplysningerData mottatteOpplysningerData) {
        List<ForetakUtland> foretakUtlandListe = mottatteOpplysningerData.foretakUtland;

        List<String> felterMedFeil = new ArrayList<>();
        for (int i = 0; i < foretakUtlandListe.size(); i++) {
            ForetakUtland foretakUtland = foretakUtlandListe.get(i);
            if (AdresseRegler.manglerForetakUtlandNavn(foretakUtland)) {
                felterMedFeil.add(String.format(FORETAK_UTLAND_NAVN, i));
            }
            if (AdresseRegler.manglerForetakUtlandLandkode(foretakUtland)) {
                felterMedFeil.add(String.format(FORETAK_UTLAND_LAND, i));
            }
        }
        return felterMedFeil.isEmpty() ? null
            : new Kontrollfeil(Kontroll_begrunnelser.MANGLENDE_OPPL_ANDRE_ARBEIDSFORHOLD_UTL, felterMedFeil, KontrolldataFeilType.FEIL);
    }
}

File: service/src/main/kotlin/no/nav/melosys/service/kontroll/feature/arbeidutland/kontroll/ArbeidUtlandKontroll.kt
package no.nav.melosys.service.kontroll.feature.arbeidutland.kontroll

import no.nav.melosys.domain.kodeverk.begrunnelser.Fartsomrader
import no.nav.melosys.domain.kodeverk.begrunnelser.Kontroll_begrunnelser
import no.nav.melosys.domain.mottatteopplysninger.MottatteOpplysningerData
import no.nav.melosys.exception.KontrolldataFeilType
import no.nav.melosys.service.kontroll.regler.AdresseRegler
import no.nav.melosys.service.validering.Kontrollfeil


open class ArbeidUtlandKontroll {
    companion object {
        @JvmStatic
        fun arbeidsstedLandManglerFelter(mottatteOpplysningerData: MottatteOpplysningerData): Kontrollfeil? {
            val fysiskArbeidsstedListe = mottatteOpplysningerData.arbeidPaaLand.fysiskeArbeidssteder
            var erUfullstendigUtfyllt = false

            for (fysiskArbeidssted in fysiskArbeidsstedListe) {
                val manglerArbeidsstedVirksomhetsnavn = AdresseRegler.manglerArbeidsstedVirksomhetsnavn(fysiskArbeidssted)
                val manglerArbeidsstedLandkode = AdresseRegler.manglerArbeidsstedLandkode(fysiskArbeidssted)


                if (manglerArbeidsstedVirksomhetsnavn || manglerArbeidsstedLandkode) {
                    erUfullstendigUtfyllt = true
                }
            }
            return if (erUfullstendigUtfyllt) Kontrollfeil(Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSSTED_LAND, KontrolldataFeilType.FEIL)
            else null
        }

        @JvmStatic
        fun maritimtArbeidsstedManglerFelter(mottatteOpplysningerData: MottatteOpplysningerData): Kontrollfeil? {
            val maritimtArbeidListe = mottatteOpplysningerData.maritimtArbeid

            if (maritimtArbeidListe.isEmpty()) {
                return null
            }

            var erUfullstendigUtfyllt = false
            for (maritimtArbeid in maritimtArbeidListe) {

                val manglerArbeidsstedVirksomhetsnavnInnretningEnhetNavn =
                    AdresseRegler.manglerArbeidsstedVirksomhetsnavnInnretningEnhetNavn(maritimtArbeid)
                val manglerArbeidsstedVirksomhetsnavnInnretningLandsTerritorialFarvann =
                    AdresseRegler.manglerArbeidsstedVirksomhetsnavnInnretningLandsTerritorialFarvann(maritimtArbeid)
                val manglerArbeidsstedVirksomhetsnavnInnretningFlaggstat =
                    AdresseRegler.manglerArbeidsstedVirksomhetsnavnInnretningFlaggstat(maritimtArbeid)

                val erSkip = maritimtArbeid.fartsomradeKode != null

                if(erSkip) {
                    if (maritimtArbeid.fartsomradeKode == Fartsomrader.INNENRIKS) {
                        if (manglerArbeidsstedVirksomhetsnavnInnretningEnhetNavn || manglerArbeidsstedVirksomhetsnavnInnretningLandsTerritorialFarvann) {
                            erUfullstendigUtfyllt = true
                        }
                    } else {
                        if (manglerArbeidsstedVirksomhetsnavnInnretningEnhetNavn || manglerArbeidsstedVirksomhetsnavnInnretningFlaggstat) {
                            erUfullstendigUtfyllt = true
                        }
                    }
                }
            }

            return if (erUfullstendigUtfyllt) Kontrollfeil(Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSSTED_MARITIM, KontrolldataFeilType.FEIL)
            else null
        }

        @JvmStatic
        fun offshoreArbeidsstedManglerFelter(mottatteOpplysningerData: MottatteOpplysningerData): Kontrollfeil? {
            val maritimtArbeidListe = mottatteOpplysningerData.maritimtArbeid

            if (maritimtArbeidListe.isEmpty()) {
                return null
            }

            var erUfullstendigUtfyllt = false
            for (maritimtArbeid in maritimtArbeidListe) {

                val manglerArbeidsstedVirksomhetsnavnInnretningEnhetNavn =
                    AdresseRegler.manglerArbeidsstedVirksomhetsnavnInnretningEnhetNavn(maritimtArbeid)
                val manglerArbeidsstedVirksomhetsnavnInnretningLandssokkel =
                    AdresseRegler.manglerArbeidsstedVirksomhetsnavnInnretningLandssokkel(maritimtArbeid)

                val erSkip = maritimtArbeid.fartsomradeKode != null

                if (!erSkip) {
                    if (manglerArbeidsstedVirksomhetsnavnInnretningEnhetNavn || manglerArbeidsstedVirksomhetsnavnInnretningLandssokkel) {
                        erUfullstendigUtfyllt = true
                    }
                }
            }

            return if (erUfullstendigUtfyllt) Kontrollfeil(Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSSTED_OFFSHORE, KontrolldataFeilType.FEIL)
            else null
        }



        @JvmStatic
        fun luftfartArbeidsstedManglerFelter(mottatteOpplysningerData: MottatteOpplysningerData): Kontrollfeil? {
            val luftfartBaseListe = mottatteOpplysningerData.luftfartBaser

            if (luftfartBaseListe.isEmpty()) {
                return null
            }

            var erUfullstendigUtfyllt = false

            for (luftfartBase in luftfartBaseListe) {
                val manglerArbeidsstedVirksomhetsnavnLuftfartBaseHjemmebaseNavn =
                    AdresseRegler.manglerArbeidsstedVirksomhetsnavnLuftfartBaseHjemmebaseNavn(luftfartBase)
                val manglerArbeidsstedVirksomhetsnavnLuftfartBaseHjemmebaseLand =
                    AdresseRegler.manglerArbeidsstedVirksomhetsnavnLuftfartBaseHjemmebaseLand(luftfartBase)

                if (manglerArbeidsstedVirksomhetsnavnLuftfartBaseHjemmebaseNavn || manglerArbeidsstedVirksomhetsnavnLuftfartBaseHjemmebaseLand) {
                    erUfullstendigUtfyllt = true
                }
            }
            return if (erUfullstendigUtfyllt) Kontrollfeil(Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSSTED_LUFTFART, KontrolldataFeilType.FEIL)
            else null
        }

        @JvmStatic
        fun foretakUtlandManglerFelter(mottatteOpplysningerData: MottatteOpplysningerData): Kontrollfeil? {
            val foretakUtlandListe = mottatteOpplysningerData.foretakUtland

            var erUfullstendigUtfyllt = false

            for (foretakUtland in foretakUtlandListe) {
                val manglerForetakUtlandNavn = AdresseRegler.manglerForetakUtlandNavn(foretakUtland)
                val manglerForetakUtlandLandkode = AdresseRegler.manglerForetakUtlandLandkode(foretakUtland)
                val manglerForetakUtlandPoststed = AdresseRegler.manglerForetakUtlandPoststed(foretakUtland)
                val manglerForetakUtlandPostnummer = AdresseRegler.manglerForetakUtlandPostnummer(foretakUtland)
                val manglerForetakUtlandAdresse = AdresseRegler.manglerForetakUtlandAdresse(foretakUtland)


                if (manglerForetakUtlandNavn || manglerForetakUtlandLandkode || manglerForetakUtlandPoststed || manglerForetakUtlandPostnummer || manglerForetakUtlandAdresse) {
                    if(!foretakUtland.selvstendigNæringsvirksomhet) {
                        erUfullstendigUtfyllt = true
                    }
                }
            }

            return if (erUfullstendigUtfyllt) Kontrollfeil(Kontroll_begrunnelser.MANGLENDE_OPPL_ARBEIDSFORHOLD_UTL, KontrolldataFeilType.FEIL)
            else null
        }

        @JvmStatic
        fun selvstendigUtlandManglerFelter(mottatteOpplysningerData: MottatteOpplysningerData): Kontrollfeil? {
            val foretakUtlandListe = mottatteOpplysningerData.foretakUtland

            var erUfullstendigUtfyllt = false

            for (foretakUtland in foretakUtlandListe) {
                val manglerForetakUtlandNavn = AdresseRegler.manglerForetakUtlandNavn(foretakUtland)
                val manglerForetakUtlandLandkode = AdresseRegler.manglerForetakUtlandLandkode(foretakUtland)
                val manglerForetakUtlandPoststed = AdresseRegler.manglerForetakUtlandPoststed(foretakUtland)
                val manglerForetakUtlandPostnummer = AdresseRegler.manglerForetakUtlandPostnummer(foretakUtland)
                val manglerForetakUtlandAdresse = AdresseRegler.manglerForetakUtlandAdresse(foretakUtland)


                if (manglerForetakUtlandNavn || manglerForetakUtlandLandkode || manglerForetakUtlandPoststed || manglerForetakUtlandPostnummer || manglerForetakUtlandAdresse) {
                    if(foretakUtland.selvstendigNæringsvirksomhet) {
                        erUfullstendigUtfyllt = true
                    }
                }
            }

            return if (erUfullstendigUtfyllt) Kontrollfeil(Kontroll_begrunnelser.MANGLENDE_OPPL_SELVSTENDIG_ARBEIDSFORHOLD_UTL, KontrolldataFeilType.FEIL)
            else null
        }
    }
}
